tabularvars.go

```diff
@@ -0,0 +1,175 @@
+	f, err := os.Open(pathToCSV)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := csv.NewReader(f)
+	records, err := r.ReadAll()
+	if err != nil {
+		return err
+	}
+
+	// Check if the variable exists in RF
+	var existingGenID int
+	generators, err := api.GetGenerators()
+	if err != nil {
+		return err
+	}
+
+	for _, gen := range generators {
+		if gen.Name == name {
+			existingGenID = gen.ID
+		}
+	}
+
+	if existingGenID != 0 {
+		if overwrite {
+			// if variable exists and we want to override it with new one delete it here
+			log.Printf("Tabular var %v exists, overwriting it with new data.\n", name)
+			err := api.DeleteGenerator(existingGenID)
+			if err != nil {
+				return err
+			}
+		} else {
+			// if variable exists but we didn't specify to override it then return with error
+			return errors.New("Tabular variable: " + name +
+				" already exists, use different name or choose an option to override it")
+		}
+	}
+
+	// prepare input data
+	columnNames, rows := records[0], records[1:]
+	parsedColumnNames := make([]string, len(columnNames))
+	for i, colName := range columnNames {
+		formattedColName := strings.TrimSpace(strings.ToLower(colName))
+		parsedColName := strings.Replace(formattedColName, " ", "_", -1)
+		parsedColumnNames[i] = parsedColName
+	}
+
+	// create new generator for the tabular variable
+	description := "Variable " + name + " uploded through cli client."
+	newGenerator, err := api.CreateTabularVar(name, description, parsedColumnNames, singleUse)
+	if err != nil {
+		return err
+	}
+
+	// batch the rows and put them into a channel
+	numOfBatches := int(math.Ceil(float64(len(rows)) / float64(tabularBatchSize)))
+	rowsToUpload := make(chan [][]string, numOfBatches)
+	for i := 0; i < len(rows); i += tabularBatchSize {
+		batch := rows[i:min(i+tabularBatchSize, len(rows))]
+		rowsToUpload <- batch
+	}
+	close(rowsToUpload)
+
+	// chan to gather errors from workers
+	errors := make(chan error, numOfBatches)
+
+	log.Println("Beginning batch upload of csv file...")
+
+	// spawn workers to upload the rows
+	for i := 0; i < tabularConcurency; i++ {
+		go rowUploadWorker(api, newGenerator, parsedColumnNames, rowsToUpload, errors)
+	}
+
+	for i := 0; i < numOfBatches; i++ {
+		if err := <-errors; err != nil {
+			return err
+		}
+		log.Printf("Tabular variable '%v' batch %v of %v uploaded.", name, i+1, numOfBatches)
+	}
+
+	return nil
+}
+
+// Well... yeah...
+func min(a, b int) int {
+	if a <= b {
+		return a
+	}
+	return b
+}
+
+// rowUploadWorker is a helper worker which reads batch of rows to upload from rows chan
+// and pushes potential errors through errorsChan
+func rowUploadWorker(api tabularVariablesAPI, generator rainforest.Generator,
+	columns []string, rowsChan <-chan [][]string, errorsChan chan<- error) {
+	for rows := range rowsChan {
+		error := api.AddGeneratorRowsFromTable(generator, columns, rows)
+		errorsChan <- error
```

Comment:
As decided on slack, we're leaving it as is, because they'll get killed anyways when main routine exits.